# What drives the price of a car?

![](images/kurt.jpeg)

**OVERVIEW**

In this application, you will explore a dataset from Kaggle. The original dataset contained information on 3 million used cars. The provided dataset contains information on 426K cars to ensure speed of processing.  Your goal is to understand what factors make a car more or less expensive.  As a result of your analysis, you should provide clear recommendations to your client -- a used car dealership -- as to what consumers value in a used car.

### CRISP-DM Framework

<center>
    <img src = images/crisp.png width = 50%/>
</center>


To frame the task, throughout our practical applications, we will refer back to a standard process in industry for data projects called CRISP-DM.  This process provides a framework for working through a data problem.  Your first step in this application will be to read through a brief overview of CRISP-DM [here](https://mo-pcco.s3.us-east-1.amazonaws.com/BH-PCMLAI/module_11/readings_starter.zip).  After reading the overview, answer the questions below.

### Business Understanding

From a business perspective, we are tasked with identifying key drivers for used car prices.  In the CRISP-DM overview, we are asked to convert this business framing to a data problem definition.  Using a few sentences, reframe the task as a data task with the appropriate technical vocabulary.

### Data Task Definition

The goal is to build a supervised regression model to find the relationship between the price of a used car and its features. The target variable is the price of the car, and the features include various attributes such as make, model, year, mileage, and condition. The goal is to identify which features are most influential in determining the price of a used car and to provide actionable insights for the used car dealership to optimize their inventory and pricing strategies.

This task involves data preprocessing, feature engineering, and multiple model building using regression techniques. The models' coefficients will be compared to identify the most significant drivers of used car prices. The performance of the models will be evaluated using appropriate metrics such as Mean Absolute Error (MAE) or Root Mean Squared Error (RMSE) to ensure that the predictions are accurate and reliable for business decision making.


### Data Understanding

After considering the business understanding, we want to get familiar with our data.  Write down some steps that you would take to get to know the dataset and identify any quality issues within.  Take time to get to know the dataset and explore what information it contains and how this could be used to inform your business understanding.

#### Steps for Data Understanding
1. Load the dataset and perform an initial inspection to understand its structure, the number of rows and columns, data types, and a preview the first and last records.
2. Check for missing values and determine how to handle them, fill them with mean, mode, or delete them.
3. Explore the distribution of the target variable (price) and identify any outliers.
4. Identify any categorical variables and explore unique values.
5. Identify any data quality issues, such as duplicates, inconsistency, or errors, and plan how to address them.

In [34]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import plotly.express as px
import seaborn as sns
from bleach import clean

from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error
from sklearn.metrics import mean_absolute_error
from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import StandardScaler

# to suppress warnings
from warnings import filterwarnings
filterwarnings('ignore')

In [35]:
df = pd.read_csv('data/vehicles.csv')

df.shape

(426880, 18)

***Interpretation:*** The dataset contains 426,000 rows and 20 columns. This indicates that we have a large dataset to work with, which can provide valuable insights into the factors that influence used car prices.

In [36]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 426880 entries, 0 to 426879
Data columns (total 18 columns):
 #   Column        Non-Null Count   Dtype  
---  ------        --------------   -----  
 0   id            426880 non-null  int64  
 1   region        426880 non-null  object 
 2   price         426880 non-null  int64  
 3   year          425675 non-null  float64
 4   manufacturer  409234 non-null  object 
 5   model         421603 non-null  object 
 6   condition     252776 non-null  object 
 7   cylinders     249202 non-null  object 
 8   fuel          423867 non-null  object 
 9   odometer      422480 non-null  float64
 10  title_status  418638 non-null  object 
 11  transmission  424324 non-null  object 
 12  VIN           265838 non-null  object 
 13  drive         296313 non-null  object 
 14  size          120519 non-null  object 
 15  type          334022 non-null  object 
 16  paint_color   296677 non-null  object 
 17  state         426880 non-null  object 
dtypes: f

In [37]:
df.head(20)

id                  region  price  year manufacturer model  \
0   7222695916                prescott   6000   NaN          NaN   NaN   
1   7218891961            fayetteville  11900   NaN          NaN   NaN   
2   7221797935            florida keys  21000   NaN          NaN   NaN   
3   7222270760  worcester / central MA   1500   NaN          NaN   NaN   
4   7210384030              greensboro   4900   NaN          NaN   NaN   
5   7222379453           hudson valley   1600   NaN          NaN   NaN   
6   7221952215           hudson valley   1000   NaN          NaN   NaN   
7   7220195662           hudson valley  15995   NaN          NaN   NaN   
8   7209064557         medford-ashland   5000   NaN          NaN   NaN   
9   7219485069                    erie   3000   NaN          NaN   NaN   
10  7218893038                 el paso      0   NaN          NaN   NaN   
11  7218325704                 el paso      0   NaN          NaN   NaN   
12  7217788283                 el paso      0   NaN          NaN   NaN   
13  7217147606                 el paso      0   NaN          NaN   NaN   
14  7209027818                 el paso      0   NaN          NaN   NaN   
15  7223509794              bellingham  13995   NaN          NaN   NaN   
16  7222753076              bellingham  24999   NaN          NaN   NaN   
17  7222206015              bellingham  21850   NaN          NaN   NaN   
18  7220030122              bellingham  26850   NaN          NaN   NaN   
19  7218423006              bellingham  11999   NaN          NaN   NaN   

   condition cylinders fuel  odometer title_status transmission  VIN drive  \
0        NaN       NaN  NaN       NaN          NaN          NaN  NaN   NaN   
1        NaN       NaN  NaN       NaN          NaN          NaN  NaN   NaN   
2        NaN       NaN  NaN       NaN          NaN          NaN  NaN   NaN   
3        NaN       NaN  NaN       NaN          NaN          NaN  NaN   NaN   
4        NaN       NaN  NaN       NaN          NaN          NaN  NaN   NaN   
5        NaN       NaN  NaN       NaN          NaN          NaN  NaN   NaN   
6        NaN       NaN  NaN       NaN          NaN          NaN  NaN   NaN   
7        NaN       NaN  NaN       NaN          NaN          NaN  NaN   NaN   
8        NaN       NaN  NaN       NaN          NaN          NaN  NaN   NaN   
9        NaN       NaN  NaN       NaN          NaN          NaN  NaN   NaN   
10       NaN       NaN  NaN       NaN          NaN          NaN  NaN   NaN   
11       NaN       NaN  NaN       NaN          NaN          NaN  NaN   NaN   
12       NaN       NaN  NaN       NaN          NaN          NaN  NaN   NaN   
13       NaN       NaN  NaN       NaN          NaN          NaN  NaN   NaN   
14       NaN       NaN  NaN       NaN          NaN          NaN  NaN   NaN   
15       NaN       NaN  NaN       NaN          NaN          NaN  NaN   NaN   
16       NaN       NaN  NaN       NaN          NaN          NaN  NaN   NaN   
17       NaN       NaN  NaN       NaN          NaN          NaN  NaN   NaN   
18       NaN       NaN  NaN       NaN          NaN          NaN  NaN   NaN   
19       NaN       NaN  NaN       NaN          NaN          NaN  NaN   NaN   

   size type paint_color state  
0   NaN  NaN         NaN    az  
1   NaN  NaN         NaN    ar  
2   NaN  NaN         NaN    fl  
3   NaN  NaN         NaN    ma  
4   NaN  NaN         NaN    nc  
5   NaN  NaN         NaN    ny  
6   NaN  NaN         NaN    ny  
7   NaN  NaN         NaN    ny  
8   NaN  NaN         NaN    or  
9   NaN  NaN         NaN    pa  
10  NaN  NaN         NaN    tx  
11  NaN  NaN         NaN    tx  
12  NaN  NaN         NaN    tx  
13  NaN  NaN         NaN    tx  
14  NaN  NaN         NaN    tx  
15  NaN  NaN         NaN    wa  
16  NaN  NaN         NaN    wa  
17  NaN  NaN         NaN    wa  
18  NaN  NaN         NaN    wa  
19  NaN  NaN         NaN    wa

In [38]:
df.tail()

id   region  price    year manufacturer  \
426875  7301591192  wyoming  23590  2019.0       nissan   
426876  7301591187  wyoming  30590  2020.0        volvo   
426877  7301591147  wyoming  34990  2020.0     cadillac   
426878  7301591140  wyoming  28990  2018.0        lexus   
426879  7301591129  wyoming  30590  2019.0          bmw   

                           model condition    cylinders    fuel  odometer  \
426875         maxima s sedan 4d      good  6 cylinders     gas   32226.0   
426876  s60 t5 momentum sedan 4d      good          NaN     gas   12029.0   
426877          xt4 sport suv 4d      good          NaN  diesel    4174.0   
426878           es 350 sedan 4d      good  6 cylinders     gas   30112.0   
426879  4 series 430i gran coupe      good          NaN     gas   22716.0   

       title_status transmission                VIN drive size       type  \
426875        clean        other  1N4AA6AV6KC367801   fwd  NaN      sedan   
426876        clean        other  7JR102FKXLG042696   fwd  NaN      sedan   
426877        clean        other  1GYFZFR46LF088296   NaN  NaN  hatchback   
426878        clean        other  58ABK1GG4JU103853   fwd  NaN      sedan   
426879        clean        other  WBA4J1C58KBM14708   rwd  NaN      coupe   

       paint_color state  
426875         NaN    wy  
426876         red    wy  
426877       white    wy  
426878      silver    wy  
426879         NaN    wy

In [39]:
df.describe()

id         price           year      odometer
count  4.268800e+05  4.268800e+05  425675.000000  4.224800e+05
mean   7.311487e+09  7.519903e+04    2011.235191  9.804333e+04
std    4.473170e+06  1.218228e+07       9.452120  2.138815e+05
min    7.207408e+09  0.000000e+00    1900.000000  0.000000e+00
25%    7.308143e+09  5.900000e+03    2008.000000  3.770400e+04
50%    7.312621e+09  1.395000e+04    2013.000000  8.554800e+04
75%    7.315254e+09  2.648575e+04    2017.000000  1.335425e+05
max    7.317101e+09  3.736929e+09    2022.000000  1.000000e+07

In [40]:
# Function that takes a dataframe a returns a new dataframe with the total of nulls and percentage of nulls for each column in dataframe
def calculate_percentage_of_nulls_by_column(dataframe):
    null_total = dataframe.isnull().sum().sort_values(ascending=False)

    null_percentage = (dataframe.isnull().sum() * 100/ dataframe.isnull().count()).round(2).sort_values(ascending=False)
    null_df = pd.DataFrame({'Total Nulls': null_total, 'Percentage of Nulls (%)': null_percentage})
    return null_df

In [41]:
calculate_percentage_of_nulls_by_column(df)

Total Nulls  Percentage of Nulls (%)
size               306361                    71.77
cylinders          177678                    41.62
condition          174104                    40.79
VIN                161042                    37.73
drive              130567                    30.59
paint_color        130203                    30.50
type                92858                    21.75
manufacturer        17646                     4.13
title_status         8242                     1.93
model                5277                     1.24
odometer             4400                     1.03
fuel                 3013                     0.71
transmission         2556                     0.60
year                 1205                     0.28
id                      0                     0.00
region                  0                     0.00
price                   0                     0.00
state                   0                     0.00

##### What would be the size of the data if I drop all missing entries?

In [42]:
non_null_df = df.dropna()
non_null_df.shape

(34868, 18)

In [43]:
# Check for unique values in categorical columns
categorical_cols = df.select_dtypes(include=['object']).columns
for col in categorical_cols:
    print(f"{col}: {df[col].nunique()} unique values")

region: 404 unique values
manufacturer: 42 unique values
model: 29649 unique values
condition: 6 unique values
cylinders: 8 unique values
fuel: 5 unique values
title_status: 6 unique values
transmission: 3 unique values
VIN: 118246 unique values
drive: 3 unique values
size: 4 unique values
type: 13 unique values
paint_color: 12 unique values
state: 51 unique values


In [44]:
# Print unique values for each column
for col in categorical_cols:
    print(f"{col}: {df[col].unique()}")  # Print first 10 unique values for each categorical column


region: ['prescott' 'fayetteville' 'florida keys' 'worcester / central MA'
 'greensboro' 'hudson valley' 'medford-ashland' 'erie' 'el paso'
 'bellingham' 'skagit / island / SJI' 'la crosse' 'auburn' 'birmingham'
 'dothan' 'florence / muscle shoals' 'gadsden-anniston'
 'huntsville / decatur' 'mobile' 'montgomery' 'tuscaloosa'
 'anchorage / mat-su' 'fairbanks' 'kenai peninsula' 'southeast alaska'
 'flagstaff / sedona' 'mohave county' 'phoenix' 'show low' 'sierra vista'
 'tucson' 'yuma' 'fort smith' 'jonesboro' 'little rock' 'texarkana'
 'bakersfield' 'chico' 'fresno / madera' 'gold country' 'hanford-corcoran'
 'humboldt county' 'imperial county' 'inland empire' 'los angeles'
 'mendocino county' 'merced' 'modesto' 'monterey bay' 'orange county'
 'palm springs' 'redding' 'reno / tahoe' 'sacramento' 'san diego'
 'san luis obispo' 'santa barbara' 'santa maria' 'SF bay area'
 'siskiyou county' 'stockton' 'susanville' 'ventura county'
 'visalia-tulare' 'yuba-sutter' 'boulder' 'colorado springs

***Interpretation:*** Above, I calculated the total number of null values and their percentage for each column in the dataset. The resulting dataframe is sorted by the total number of nulls in descending order, the variable with the highest number of nulls will appear first.

If I drop all missing entries, the resulting dataframe will have `34,868` rows, a significant reduction from the original `426,880` rows. First, I will consider removing the unnecessary columns, and then try again removing the missing entries to validate the amount of data we have left. Based on that result, I will decide which is a better approach if removing the missing entries or fill the data with the mode, mean, or any other value.

The `size` column has the highest number of null values, missing `71.8%` of its entries, which indicates that it may not be a reliable feature for modeling. The `cylinders` column is missing `41.6%` of its entries, which is also a significant amount of missing data. The `condition` column is missing `40.79%` of entries, also a significant amount. The `drive` and `paint_color` columns are missing `30.5%` of entries. The `type` column is missing `21.8%` of entries. The other columns are missing less than `5%` of their entries, which may be more manageable.

I will handle the missing data as follows:
- I will drop the `size` column, as it has the highest number of null values, missing `71.8%` of its entries, which indicates that it may not be a reliable feature for modeling.
- For `cylinders`, since is missing `41.6%` of values, if I fill the data with the mode (most common value), it could reduce the variance of the data and cause potential bias. This is a categorical variable, one of the possible values is `other`, I will assign `other` to all missing entries.
- The `condition` column is also missing significant amount of entries `40.79%`.  This is also a categorical column, I will create a new categorical value called `unknown` and fill the missing values with it. The model will learn if the absence of this information has an impact on price.
- For `drive`, it's missing a significant amount of entries `30.6%`. This is also a categorical column, I will create a new categorical value called `unknown` and fill the missing values with it. The model will learn if the absence of this information has an impact on price.
- For `paint_color` column, the value varies by brand and year. I will create new categorical value called `unknown` and fill the missing values with it.
- For `type`, I will also consider filling the missing values with value `other`, which is a value already present in the column.
- For other columns with less than `5%` missing values, I will consider removing the rows with missing values.

In [45]:
counts, bin_edges = np.histogram(df['price'], bins=50)
bin_centers = (bin_edges[:-1] + bin_edges[1:]) / 2

agg_df = pd.DataFrame({"bin_center": bin_centers, "count": counts})
fig = px.bar(agg_df, x="bin_center", y="count")
fig.update_layout(title='Distribution of Car Prices')
fig.show()

***Interpretation:*** The distribution is not clearly seen because there are some extreme outliers in the price variable. To better visualize the distribution, I will apply a logarithmic transformation to the price variable and then plot the histogram again.

In [46]:
prices = np.log1p(df["price"]).to_numpy()
counts, bin_edges = np.histogram(prices, bins=50)
bin_centers = (bin_edges[:-1] + bin_edges[1:]) / 2

agg_df = pd.DataFrame({"price": bin_centers, "count": counts})
fig = px.bar(agg_df, x="price", y="count")
fig.update_layout(title_text='Distribution of Car Prices (Log Transformed)')
fig.show()

***Interpretation***: The price in this data has some extreme outliers. The log transformation helps to reduce the impact of these outliers and provides a more normal distribution of the price variable. However, I will remove the outliers using the IQR method.

### Data Preparation

After our initial exploration and fine-tuning of the business understanding, it is time to construct our final dataset prior to modeling.  Here, we want to make sure to handle any integrity issues and cleaning, the engineering of new features, any transformations that we believe should happen (scaling, logarithms, normalization, etc.), and general preparation for modeling with `sklearn`.

In [47]:
# Clean data
# drop unusable columns: size (many missing values), VIN (unique id, not useful), id (inventory id, not useful)
clean_df = df.drop(columns=['size', 'VIN', 'id'])
clean_df

region  price    year manufacturer  \
0                     prescott   6000     NaN          NaN   
1                 fayetteville  11900     NaN          NaN   
2                 florida keys  21000     NaN          NaN   
3       worcester / central MA   1500     NaN          NaN   
4                   greensboro   4900     NaN          NaN   
...                        ...    ...     ...          ...   
426875                 wyoming  23590  2019.0       nissan   
426876                 wyoming  30590  2020.0        volvo   
426877                 wyoming  34990  2020.0     cadillac   
426878                 wyoming  28990  2018.0        lexus   
426879                 wyoming  30590  2019.0          bmw   

                           model condition    cylinders    fuel  odometer  \
0                            NaN       NaN          NaN     NaN       NaN   
1                            NaN       NaN          NaN     NaN       NaN   
2                            NaN       NaN          NaN     NaN       NaN   
3                            NaN       NaN          NaN     NaN       NaN   
4                            NaN       NaN          NaN     NaN       NaN   
...                          ...       ...          ...     ...       ...   
426875         maxima s sedan 4d      good  6 cylinders     gas   32226.0   
426876  s60 t5 momentum sedan 4d      good          NaN     gas   12029.0   
426877          xt4 sport suv 4d      good          NaN  diesel    4174.0   
426878           es 350 sedan 4d      good  6 cylinders     gas   30112.0   
426879  4 series 430i gran coupe      good          NaN     gas   22716.0   

       title_status transmission drive       type paint_color state  
0               NaN          NaN   NaN        NaN         NaN    az  
1               NaN          NaN   NaN        NaN         NaN    ar  
2               NaN          NaN   NaN        NaN         NaN    fl  
3               NaN          NaN   NaN        NaN         NaN    ma  
4               NaN          NaN   NaN        NaN         NaN    nc  
...             ...          ...   ...        ...         ...   ...  
426875        clean        other   fwd      sedan         NaN    wy  
426876        clean        other   fwd      sedan         red    wy  
426877        clean        other   NaN  hatchback       white    wy  
426878        clean        other   fwd      sedan      silver    wy  
426879        clean        other   rwd      coupe         NaN    wy  

[426880 rows x 15 columns]

In [48]:
# Remove outliers using IQR method
Q1 = clean_df['price'].quantile(0.25)
Q3 = clean_df['price'].quantile(0.75)
IQR = Q3 - Q1

# Remove outliers
clean_df = clean_df[(clean_df['price'] >= (Q1 - 1.5 * IQR)) & (clean_df['price'] <= (Q3 + 1.5 * IQR))]
clean_df

region  price    year manufacturer  \
0                     prescott   6000     NaN          NaN   
1                 fayetteville  11900     NaN          NaN   
2                 florida keys  21000     NaN          NaN   
3       worcester / central MA   1500     NaN          NaN   
4                   greensboro   4900     NaN          NaN   
...                        ...    ...     ...          ...   
426875                 wyoming  23590  2019.0       nissan   
426876                 wyoming  30590  2020.0        volvo   
426877                 wyoming  34990  2020.0     cadillac   
426878                 wyoming  28990  2018.0        lexus   
426879                 wyoming  30590  2019.0          bmw   

                           model condition    cylinders    fuel  odometer  \
0                            NaN       NaN          NaN     NaN       NaN   
1                            NaN       NaN          NaN     NaN       NaN   
2                            NaN       NaN          NaN     NaN       NaN   
3                            NaN       NaN          NaN     NaN       NaN   
4                            NaN       NaN          NaN     NaN       NaN   
...                          ...       ...          ...     ...       ...   
426875         maxima s sedan 4d      good  6 cylinders     gas   32226.0   
426876  s60 t5 momentum sedan 4d      good          NaN     gas   12029.0   
426877          xt4 sport suv 4d      good          NaN  diesel    4174.0   
426878           es 350 sedan 4d      good  6 cylinders     gas   30112.0   
426879  4 series 430i gran coupe      good          NaN     gas   22716.0   

       title_status transmission drive       type paint_color state  
0               NaN          NaN   NaN        NaN         NaN    az  
1               NaN          NaN   NaN        NaN         NaN    ar  
2               NaN          NaN   NaN        NaN         NaN    fl  
3               NaN          NaN   NaN        NaN         NaN    ma  
4               NaN          NaN   NaN        NaN         NaN    nc  
...             ...          ...   ...        ...         ...   ...  
426875        clean        other   fwd      sedan         NaN    wy  
426876        clean        other   fwd      sedan         red    wy  
426877        clean        other   NaN  hatchback       white    wy  
426878        clean        other   fwd      sedan      silver    wy  
426879        clean        other   rwd      coupe         NaN    wy  

[418703 rows x 15 columns]

In [49]:
calculate_percentage_of_nulls_by_column(clean_df)

Total Nulls  Percentage of Nulls (%)
cylinders          173763                    41.50
condition          168843                    40.33
drive              128672                    30.73
paint_color        127724                    30.50
type                91061                    21.75
manufacturer        16805                     4.01
title_status         8029                     1.92
model                4933                     1.18
odometer             4322                     1.03
fuel                 2966                     0.71
transmission         2489                     0.59
year                 1012                     0.24
region                  0                     0.00
price                   0                     0.00
state                   0                     0.00

In [50]:
# The possible values for cylinders are the following: ['8 cylinders', '6 cylinders', '4 cylinders', '5 cylinders','other', '3 cylinders', '10 cylinders', '12 cylinders']. I will assign other to all missing entries.

clean_df['cylinders'].fillna('other', inplace=True)
clean_df['cylinders'].unique()

array(['other', '8 cylinders', '6 cylinders', '4 cylinders',
       '5 cylinders', '3 cylinders', '10 cylinders', '12 cylinders'],
      dtype=object)

In [51]:
# The `condition` column will be filled with `unknown`.

clean_df['condition'].fillna('unknown', inplace=True)
clean_df['condition'].unique()

array(['unknown', 'good', 'excellent', 'fair', 'like new', 'new',
       'salvage'], dtype=object)

In [52]:
# Fill drive missing values with unknown
clean_df['drive'].fillna('unknown', inplace=True)
clean_df['drive'].unique()

array(['unknown', 'rwd', '4wd', 'fwd'], dtype=object)

In [53]:
#- For `paint_color` column, the value varies by brand and year. I will create new categorical value called `unknown` and fill the missing values with it.
clean_df['paint_color'].fillna('unknown', inplace=True)
clean_df['paint_color'].unique()

array(['unknown', 'white', 'blue', 'red', 'black', 'silver', 'grey',
       'brown', 'yellow', 'orange', 'green', 'custom', 'purple'],
      dtype=object)

In [54]:
# - For `type`,filling the missing values with value `other`
clean_df['type'].fillna('other', inplace=True)
clean_df['type'].unique()


array(['other', 'pickup', 'truck', 'coupe', 'SUV', 'hatchback',
       'mini-van', 'sedan', 'offroad', 'bus', 'van', 'convertible',
       'wagon'], dtype=object)

In [55]:
# Validate how many nan are left
calculate_percentage_of_nulls_by_column(clean_df)

Total Nulls  Percentage of Nulls (%)
manufacturer        16805                     4.01
title_status         8029                     1.92
model                4933                     1.18
odometer             4322                     1.03
fuel                 2966                     0.71
transmission         2489                     0.59
year                 1012                     0.24
region                  0                     0.00
price                   0                     0.00
condition               0                     0.00
cylinders               0                     0.00
drive                   0                     0.00
type                    0                     0.00
paint_color             0                     0.00
state                   0                     0.00

In [56]:
# - For other columns with less than `5%` missing values, I will remove the rows with missing values.
clean_df.dropna(inplace=True)
calculate_percentage_of_nulls_by_column(clean_df)

Total Nulls  Percentage of Nulls (%)
region                  0                      0.0
price                   0                      0.0
year                    0                      0.0
manufacturer            0                      0.0
model                   0                      0.0
condition               0                      0.0
cylinders               0                      0.0
fuel                    0                      0.0
odometer                0                      0.0
title_status            0                      0.0
transmission            0                      0.0
drive                   0                      0.0
type                    0                      0.0
paint_color             0                      0.0
state                   0                      0.0

***Interpretation:*** From above dataframe, I can confirm that all missing values have been fixed.

#### Plots

### Modeling

With your (almost?) final dataset in hand, it is now time to build some models.  Here, you should build a number of different regression models with the price as the target.  In building your models, you should explore different parameters and be sure to cross-validate your findings.

### Evaluation

With some modeling accomplished, we aim to reflect on what we identify as a high-quality model and what we are able to learn from this.  We should review our business objective and explore how well we can provide meaningful insight into drivers of used car prices.  Your goal now is to distill your findings and determine whether the earlier phases need revisitation and adjustment or if you have information of value to bring back to your client.

### Deployment

Now that we've settled on our models and findings, it is time to deliver the information to the client.  You should organize your work as a basic report that details your primary findings.  Keep in mind that your audience is a group of used car dealers interested in fine-tuning their inventory.